s3/src/main/java/com/instaclustr/kafka/connect/s3/RecordFormat0.java

```
@@ -1,126 +1,89 @@
+    public SourceRecord readRecord(final String jsonRow, final Map<String, ?> sourcePartition,
+                                   final Map<String, Object> sourceOffset, final String topic, final int partition) throws IOException, NumberFormatException {
+
+        try {
+            JsonObject jsonObject = jsonParser.parse(jsonRow).getAsJsonObject();
+
+            if (jsonObject.isJsonObject()) {
+                byte[] key = (jsonObject.get("k").isJsonNull()) ? null : readAsObjectOrString(jsonObject, "k").getBytes(StandardCharsets.UTF_8);
+                byte[] value = (jsonObject.get("v").isJsonNull()) ? null : readAsObjectOrString(jsonObject, "v").getBytes(StandardCharsets.UTF_8);
+                long timestamp = jsonObject.get("t").getAsLong();
+                long offset = jsonObject.get("o").getAsLong();
+
+                sourceOffset.put("lastReadOffset", offset);
+                return new SourceRecord(sourcePartition, sourceOffset, topic, partition, Schema.BYTES_SCHEMA, key, Schema.BYTES_SCHEMA, value, timestamp);
             } else {
-                headers.add(key, this.headerConverter.toConnectHeader(topic, key, dataInputStream.readNBytes(headerValueLength)));
+                logger.error("Did not receive a json object " + jsonRow);
+                throw new IOException("Did not receive a json object " + jsonRow);
             }
+        } catch (Exception e) {
+            logger.error("Could not construct Source Record, reason: " + e.getMessage());
+            throw new IOException("Could not construct Source Record, reason: " + e.getMessage());
+        }
+    }
+
+    private String constructJson(Record record) {
+        return gson.toJson(record);
+    }
+
+    private String readAsObjectOrString(JsonObject jsonObject, String memberName) {
+        try {
+            return jsonObject.getAsJsonObject(memberName).toString();
+        } catch (ClassCastException e) {
+            return jsonObject.get(memberName).getAsString();
```

Using exceptions as way of control flow in a program is risky as it can en-maskarate real problems, and cause silent failures, In my oppinion, we should have two functions, one for keys and the other for values and consider ClassCastException as fatal errors.

---

> Using exceptions as way of control flow in a program is risky

I agree with that. Not sure how we can tackle this. Open for suggestions.

> we should have two functions, one for keys and the other for values

That would lead us to duplicate code cause both keys and values can be either Strings of Json values.